### DISCUSSION TITLE
State of window recording

### DISCUSSION THREAD

**Discussion by oct2pus** [labwc/labwc#2354] [Q&A] [answered]
Hi

I draw art and I tend to record my work window, I can currently work around the limitations by simply fullscreening Krita, but its obviously not ideal. I had figured [foreign-toplevel: add ext-toplevel-list](https://github.com/labwc/labwc/commit/d6a48ab7a7cc3fbf01cf654ff19b38b4a737dc6f) would fix this, but after changing to the git version this does not seem to be the case.

I did try and do some reading, I assumed [this merge request](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/124) was the blocker, but it seems to have been merged in. I'm not really finding any good resources of what I need to keep an eye on.

What are the current blockers for streaming individual windows, so that I can keep an eye on it.

I've also noticed discord screen recording seems to be broken, (the preview window works, but discord does not display any video), this isn't as necessary to me but it'd also be a nice to have.

**Comment by Consolatis** [comment, marked as answer]
> I did try and do some reading, I assumed [this merge request](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/124) was the blocker, but it seems to have been merged in. I'm not really finding any good resources of what I need to keep an eye on.

The protocol you found is the right one. Now it needs
- an implementation on the wlroots side which is [this MR](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4545) (note that it may not actually implement window capture but only output capture using the new protocol, I didn't follow its development for a while)
- a wlroots release using the new protocol and support for window capture
- labwc using that wlroots version

All in all I don't expect window capture to work on labwc for at least half a year, potentially longer depending on the wlroots MR linked above and if it also implements window capture or not.

Edit:
> I've also noticed discord screen recording seems to be broken, (the preview window works, but discord does not display any video), this isn't as necessary to me but it'd also be a nice to have.

I can't really say anything about that, maybe someone else has an idea what is going on there.

**Reply by oct2pus** [reply]
I appreciate the info! I'm glad everything is at least ready to go on the labwc side. My current streaming needs doesn't really need more than the actual krita window itself. If this ever becomes a real problem I can always swap back to GNOME/KDE and come back to labwc later (unfortunate but at least I have options!)

**Reply by johanmalm** [reply]
I don't think it would be that hard to do a temporary patch on `action.c` (here: https://github.com/labwc/labwc/blob/f8ed199197dd68de3267f5c1429b481f60f4ef72/src/action.c#L824) to add an action to execute grim on a particular window. I think it would feel a bit too "hacky" to upstream, but might be useful to just get a particular job done.

**Reply by oct2pus** [reply]
I discovered a workaround (just for obs) that's less invasive than maintaining a private fork of labwc.

https://github.com/nowrep/obs-vkcapture

still less than ideal (need to keep a terminal open) but i can stream just my window capture now, seems to work with krita so no more accidental web browser/file browser/discord sharing.

**Reply by Consolatis** [reply]
Are you running Krita via xwayland? If that is the case any X11 screencapture application might work as well (didn't verify though).

**Reply by oct2pus** [reply]
@Consolatis yeah. currently krita only runs on x11 and i don't suspect we'll see any changes until a qt6 port occurs (lots of custom QT stuff being used).

You are correct tho! I did force obs to launch via x11 and I can also record my Krita window that way. I'm surprised I forgot to try that.
